opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py

```diff
@@ -564,6 +562,79 @@ def _format_links(links):
         return f_links
 
 
+class SpanLimits:
```

Comment:
I don't understand why use a class here. If the limits are to be set when creating a tracer provider, can't the limits themselves be tracer provider arguments? How does having a class will make it less ugly to support more limits? Also, everything is a class in Java, right?

Reply:
It is bundling all span limit variables and the logic to process/set-defaults together. Seems like isolated enough functionality to me. These limits can be instantiated once and passed around by provider to tracer, span and then down to individual functions used by spans internally. Without this, the logic to read/set-defaults either goes into the tracer provider (I don't think it belongs there) or to a function that gets called by the provider internally. It also means provider has to pass all individual limits as different variables down the chain meaning provider, tracer, span and internal functions all receive these 3 integers as arguments which will grow to 6 soon and in future could grow even further. 

```
Provider(max_events_limit, max_attributes_limit, max_links_limit)
     max_events_limit, max_attributes_limit, max_links_limit = _process_limit_defaults(max_events_limit, max_attributes_limit, max_links_limit)
     Resource(max_attributes_limit)
     Tracer(max_events_limit, max_attributes_limit, max_links_limit)
           Span(max_events_limit, max_attributes_limit, max_links_limit)
                 _process_links(max_links_limit)
                 _enforce_attribute(max_attributes_limit)
                 // other functions that may pass these down the chain
```

The above seems uglies to me compared to the following but I acknowledge that it is a big subjective and comes down to personal taste to some extent.

```
Provider(limits: SpanLimits)
     Resource(limits)
     Tracer(limits)
           Span(limits)
                 _process_links(limits.max_links)
                 _enforce_attribute(limits.max_attributes)
                 // other functions that may pass these down the chain
```
Note that we only support 3 of 5 limits specified by spec right now and one more will be added soon. So that increases the number of arguments to 6 in near future.

I feel like this is self-contained enough that provider shouldn't care about the logic to process these limits and instead just consume them. To me this looks more like a Sampler or IdGenerator to me. We could pass raw sampling decision or id generation functions for trace/span ID as individual arguments to provider but they are logically bound together that it makes sense to bundle them in a single type. A class also gives us a bit more flexibility in the long run. For example, limits can be dynamically updated by a remote configutation without having to re-create providers/tracers.

That said, what we need today work by passing each limit as an argument to provider/tracer/span. So happy to change if others feel too strongly about it.

BTW how about if we rename it to just `Limits` instead of `SpanLimits`? That should make it a bit more future proof as it'll allow us to use same thing for metrics and later logging as well.

Reply:
### On the class increasing API surface.

The class does indeed increase the API surface but it is relatively harmless IMO given how the class does not affect anything on it's own. It only reads values passed to it as arguments, read values from the environment and sets some variables. This means if for some reason we don't want to use this in future, we can just leave the class as is without it affecting anything else.

On the other hand, adding new arguments to tracer provider is a much bigger deal IMO. Even if don't add a new class to the API, we'd still be adding a bunch of arguments to the tracer provider constructor. I think that is probably something much more important get right.

### Alternative

One thing we could do is to neither add any new public classes, nor change the provider constructor signature. Instead we rename the class to `_SpanLimits` and add a private method on provider so it can be used like:


```python
provider = TracerProvider()  # internally sets self._limits to default _Limits
provider._set_limits(_Limits(max_events=10))

# provider = TracerProvider()._with_limits(_Limits(max_events=10))
```

This allows us to solve the issue with lazy evaluation of limits while not increasing the API surface. Distros can still us the private methods to set custom limits. This will buy us time to experiment with the design and make it public in future if desired. 

One downside is that this still doesn't allow end users to set limits via code but I think that's fine and can be solved later.

Reply:
Setting limits via code is a "Feature" that we chose to implement correct? What do you think about just having a `kwargs` in the `traceprovider` constructor?

Also, I understand the messiness of having to pass 3-5 variables down the call chain. Can we pass in variables through the constructor, but use `_SpanLimit` internally maybe?

Reply:
I wanted to bring this up in the SIG meeting yesterday but couldn't join.

Yes, we can do that but the question still remains whether we prefer a single `limits` kwargs vs `3-6` and possible an ever growing list of limit kwargs in tracer provider. It sounds like you and @ocelotl at least prefer individual kwargs over an encapsulated single kwarg.

We can go a step further for this PR and not add any new attributes to tracer provider constructor at all. Internall tracer provider will construct an instance of `_SpanLimit` and pass it down the chain. That way we will have lazy evaluation in this PR while we discuss what doing this in code should look like. May be we can live with just env var based config for a while and buy some time to think about the Python interface for this especially as the metrics API stablizes.

Reply:
I see how putting all these limits inside an object seems more convenient, but I don't see any real difference between passing the limits in an object that got them previously or in passing them directly as arguments. `kwargs` should not be used for this purpose because these arguments are known beforehand.

A couple of questions. Span Limits look like `Span` attributes. Why are we passing them to a resource, tracerprovider, tracer, etc? This feels like they should just be added in the `Span`. If we add these span limits to the arguments a tracer provider receives, why don't we also pass the rest of the arguments the `Span` receives to the tracer provider?

Maybe we can just add them to the Span arguments. If that is the case, we could just use take their values from environment variables as usual.

Reply:
I've removed ability to configure limits via code to simply the PR (https://github.com/open-telemetry/opentelemetry-python/pull/1839#issuecomment-841571686) but to answer your questions:

> Span Limits look like Span attributes. Why are we passing them to a resource, tracerprovider, tracer, etc? 

You're right. It's a bad name. I've renamed it to `_Limits`. These will be used with resources very soon and at some point with metrics.


> If we add these span limits to the arguments a tracer provider receives, why don't we also pass the rest of the arguments the Span receives to the tracer provider?

Because the feature was implement "limit configuration with code". There is no way for users to pass something to every single Span's contructor. It can only be done with manually created spans and even then usually users want limits to be configured once or once per provider (different limits for manual vs instrumentation spans or different limits for different instrumentation), not once per span.

We could have something like `_set_global_limits()` similar to global provider and propagator but that wouldn't allow limits to be set per instrumentation or different limits for different components of a service which would be really great to have. In order to make this happen, IMO, the best place to configure limits is TracerProvider (and in future MeterProvider) constructor.

That was the rationale but I've removed it all for now in order to have no Python API changes. I'm still creating a single _Limits instance per provider to not run the code over and over again for every single span or tracer and I'll be passing it to resources in an upcoming PR (again without public API changes).

Reply:
Let's discuss this during the SIG meeting. Please add this to the list of topics.